resource/repositories/repository.go

```diff
@@ -52,18 +54,24 @@ func (cfo operationsRepository) get(name string) (resource.Resource, io.ReadClos
 	return res, reader, nil
 }
 
-// set stores the resource info and data in the cache,
-// if there is one. If no cache is in use then this is a no-op. Note
-// that the returned reader may or may not be the same one that was
-// passed in.
-func (cfo operationsRepository) set(chRes charmresource.Resource, reader io.ReadCloser) (resource.Resource, io.ReadCloser, error) {
+// set stores the resource info and data in a repo, if there is one.
+// If no repo is in use then this is a no-op. Note that the returned
+// reader may or may not be the same one that was passed in.
+func (cfo operationsRepository) set(chRes charmresource.Resource, reader io.ReadCloser) (_ resource.Resource, _ io.ReadCloser, err error) {
 	if cfo.repo == nil {
 		res := resource.Resource{
 			Resource: chRes,
 		}
 		return res, reader, nil // a no-op
 	}
-	defer func() { _ = reader.Close() }()
+	defer func() {
+		if err != nil {
+			// With no err, the reader was closed down in unitSetter Read().
```

Comment:
This is an ugly hack. We should change `Read` to accept an `io.Reader` and have the caller deal with closing it (assuming it is a reader that can be closed and not a StringReader or something like that)

Reply:
The io.Reader needs to be re-worked here, it leaks out everywhere!